File: leetcode_problems/python/lc518_coin_change_2.py
class Solution:
    def change(
        self,
        amount: int,
        coins: list[int],
    ) -> int:
        n = amount + 1
        dp = [0] * n
        dp[0] = 1
        for coin in coins:
            for i in range(coin, n):
                dp[i] += dp[i - coin]
        return dp[amount]


if __name__ == "__main__":
    amount = 5
    coins = [1, 2, 5]
    print(Solution().change(amount, coins))

File: leetcode_problems/typescript/src/lc518_coin_change_2.ts
function change(amount: number, coins: number[]): number {
  let n = amount + 1;
  let dp = new Array(n).fill(0);
  dp[0] = 1;
  for (let coin of coins) {
    for (let i = coin; i < n; i++) {
      dp[i] += dp[i - coin];
    }
  }
  return dp[amount];
}
